File: mac-agent/input_injector.py
"""
input_injector.py - Injects mouse/touch events into the iOS Simulator window
using macOS CoreGraphics (CGEventPost).

The iOS Simulator maps mouse events directly to touch events, so simulating
a mouse click at the simulator window coordinates is all that's needed.
"""
import Quartz


def _to_screen_point(norm_x, norm_y, bounds):
    """Convert normalized (0–1) coordinates to absolute screen coordinates."""
    x = bounds["X"] + norm_x * bounds["Width"]
    y = bounds["Y"] + norm_y * bounds["Height"]
    return Quartz.CGPoint(x, y)


def click_at(norm_x, norm_y, bounds):
    """Simulate a tap/click at normalized coordinates within the simulator window."""
    point = _to_screen_point(norm_x, norm_y, bounds)

    for event_type in (Quartz.kCGEventLeftMouseDown, Quartz.kCGEventLeftMouseUp):
        event = Quartz.CGEventCreateMouseEvent(
            None, event_type, point, Quartz.kCGMouseButtonLeft
        )
        Quartz.CGEventPost(Quartz.kCGHIDEventTap, event)


def drag(norm_x1, norm_y1, norm_x2, norm_y2, bounds, steps=20):
    """Simulate a drag/swipe gesture between two normalized positions."""
    p1 = _to_screen_point(norm_x1, norm_y1, bounds)
    p2 = _to_screen_point(norm_x2, norm_y2, bounds)

    # Mouse down at start
    down = Quartz.CGEventCreateMouseEvent(
        None, Quartz.kCGEventLeftMouseDown, p1, Quartz.kCGMouseButtonLeft
    )
    Quartz.CGEventPost(Quartz.kCGHIDEventTap, down)

    # Drag across intermediate points
    for i in range(1, steps + 1):
        t = i / steps
        mid = Quartz.CGPoint(
            p1.x + (p2.x - p1.x) * t,
            p1.y + (p2.y - p1.y) * t,
        )
        drag_event = Quartz.CGEventCreateMouseEvent(
            None, Quartz.kCGEventLeftMouseDragged, mid, Quartz.kCGMouseButtonLeft
        )
        Quartz.CGEventPost(Quartz.kCGHIDEventTap, drag_event)

    # Mouse up at end
    up = Quartz.CGEventCreateMouseEvent(
        None, Quartz.kCGEventLeftMouseUp, p2, Quartz.kCGMouseButtonLeft
    )
    Quartz.CGEventPost(Quartz.kCGHIDEventTap, up)


def scroll_at(norm_x, norm_y, dy, bounds):
    """Simulate a scroll wheel event at normalized coordinates."""
    point = _to_screen_point(norm_x, norm_y, bounds)

    # Move cursor to position first
    move = Quartz.CGEventCreateMouseEvent(
        None, Quartz.kCGEventMouseMoved, point, Quartz.kCGMouseButtonLeft
    )
    Quartz.CGEventPost(Quartz.kCGHIDEventTap, move)

    # Scroll
    scroll = Quartz.CGEventCreateScrollWheelEvent(
        None, Quartz.kCGScrollEventUnitPixel, 1, int(dy)
    )
    Quartz.CGEventPost(Quartz.kCGHIDEventTap, scroll)

File: mac-agent/input_injector.js
/**
 * input_injector.js - Injects mouse/touch events into the iOS Simulator window
 * using macOS CoreGraphics APIs or robotjs library.
 *
 * The iOS Simulator maps mouse events directly to touch events, so simulating
 * a mouse click at the simulator window coordinates is all that's needed.
 */

import robot from 'robotjs';

// Simple delay helper
const delay = (ms) => new Promise((r) => setTimeout(r, ms));

/**
 * Convert normalized (0–1) coordinates to absolute screen coordinates.
 */
function toScreenPoint(normX, normY, bounds) {
  const x = bounds.X + normX * bounds.Width;
  const y = bounds.Y + normY * bounds.Height;
  return { x: Math.round(x), y: Math.round(y) };
}

/**
 * Simulate a tap/click at normalized coordinates within the simulator window.
 */
export function clickAt(normX, normY, bounds) {
  try {
    const point = toScreenPoint(normX, normY, bounds);
    
    // Move mouse to position
    robot.moveMouse(point.x, point.y);
    
    // Simulate click
    robot.mouseClick();
  } catch (err) {
    console.error('clickAt error:', err.message);
  }
}

/**
 * Simulate a drag/swipe gesture between two normalized positions.
 */
export async function drag(normX1, normY1, normX2, normY2, bounds, steps = 20) {
  try {
    const p1 = toScreenPoint(normX1, normY1, bounds);
    const p2 = toScreenPoint(normX2, normY2, bounds);

    // Move to start position
    robot.moveMouse(p1.x, p1.y);
    
    // Press down
    robot.mouseToggle('down');
    await delay(50);

    // Drag across intermediate points
    for (let i = 1; i <= steps; i++) {
      const t = i / steps;
      const mid = {
        x: Math.round(p1.x + (p2.x - p1.x) * t),
        y: Math.round(p1.y + (p2.y - p1.y) * t),
      };
      robot.moveMouse(mid.x, mid.y);
      await delay(5);
    }

    // Release
    robot.mouseToggle('up');
  } catch (err) {
    console.error('drag error:', err.message);
  }
}

/**
 * Simulate a scroll wheel event at normalized coordinates.
 */
export async function scrollAt(normX, normY, dy, bounds) {
  try {
    const point = toScreenPoint(normX, normY, bounds);

    // Move cursor to position
    robot.moveMouse(point.x, point.y);
    await delay(10);

    // Scroll (robotjs uses pixels; positive = up, negative = down)
    // dy is in pixels, we'll scroll in chunks
    const scrollSteps = Math.max(1, Math.abs(dy) / 20);
    const scrollDirection = dy > 0 ? 1 : -1;

    for (let i = 0; i < scrollSteps; i++) {
      robot.scroll(0, scrollDirection * 3);
      await delay(20);
    }
  } catch (err) {
    console.error('scrollAt error:', err.message);
  }
}
